subsys/modbus/modbus_core.c

```diff
@@ -137,6 +137,8 @@ void modbus_tx_adu(struct modbus_context *ctx)
 
 int modbus_tx_wait_rx_adu(struct modbus_context *ctx)
 {
+	k_sem_reset(&ctx->client_wait_sem);
```

Comment:
Would you like to add a test case to catch this in the tests?

Reply:
If I get some leeway at work, I'll get an nucleo board working and submit a test case. I think the easiest way to do it would be to artificially delay the server response past the timeout. But I'm not sure when I'll get to it.